nixos/nextcloud: use LoadCredential to read secrets

networkException opened the pull request:
This pull request adds support for using systemd's `LoadCredential=` feature to read various secret files used by nextcloud service units.

Previously credentials had to be readable by the nextcloud user, this is now no longer required.

The `nextcloud-occ` wrapper script has been adjusted to use `systemd-run` for loading credentials when being called from outside a service.

See the individual commits for more details.

Depends on #365442
Continuation of #152141

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

networkException committed:
nixos/tests/nextcloud: test for notify_push in with-declarative-redis-and-secrets

This patch adds a subtest and corresponding configuration to
with-declarative-redis-and-secrets to test for nextcloud notify_push
to be working, just as in with-postgresql-and-redis.

As notify_push needs to connect to the database, including it
in this test checks that it can read the dbpassFile properly.

networkException committed:
nixos/nextcloud: use writeShellApplication for nextcloud-occ

This patch replaces the use of writeScriptBin for the nextcloud-occ
script with writeShellApplication, enabling shell checking.

This patch also updates various invocations of the script to
use lib.getExe.

networkException committed:
nixos/nextcloud: move systemd service overrides for phpfpm-nextcloud closer to phpfpm config

networkException committed:
nixos/nextcloud: use LoadCredential to read secrets

This patch adds support for using systemd's LoadCredential
feature to read various secret files used by nextcloud service
units.

Previously credentials had to be readable by the nextcloud user,
this is now no longer required.

The nextcloud-occ wrapper script has been adjusted to use
systemd-run for loading credentials when being called from
outside a service.

In detail this change touches various details of the module:

- The nix_read_secret() php function now takes the name of a
  file relative to the path specified in the CREDENTIALS_DIRECTORY
  environment variable.
- The nix_read_secret() now exits with error code 1 instead of
  throwing a RuntimeException as this will properly error out
  the nextcloud-occ script
- Only the nextcloud-setup service unit has the adminpass credential
  added in addition to the other credentials
- Uses of ExecCondition= in nextcloud-cron and nextcloud-update-db
  have been replaced by a shell conditional as ExecCondition currently
  doesn't support credentials
- The phpfpm-nextcloud service now runs a preStart script to make
  the credentials it gets readable by the nextcloud user as the
  unit runs as root but the php process itself as nextcloud.
- To invoke occ notify_push:setup when using nextcloud notify_push
  a new service has been added that replaces the preStart script
  in nextcloud-notify_push.service. This has been done as the
  main executable only needs the database password credential.

Co-authored-by: lassulus <lassulus@lassul.us>

networkException commented:
This should be generally well tested already so reviews are encouraged, undrafting when

- [ ] I've run tests against `master`
- [ ] I've tested this on my deployment
- [ ] https://github.com/NixOS/nixpkgs/pull/365442 is merged

SuperSandro2000 commented on nixos/tests/nextcloud/with-declarative-redis-and-secrets.nix:
```diff
@@ -97,6 +104,10 @@ runTest (
           nextcloud.fail('test 0 -lt "$(redis-cli --pass secret --json KEYS "*" | jq "len")"')
+
+      with subtest("notify-push"):
+          client.execute("${pkgs.lib.getExe pkgs.nextcloud-notify_push.passthru.test_client} http://nextcloud ${config.adminuser} ${config.adminpass} >&2 &")
```
```suggestion
          client.execute("${lib.getExe pkgs.nextcloud-notify_push.passthru.test_client} http://nextcloud ${config.adminuser} ${config.adminpass} >&2 &")
```

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -1009,57 +1024,94 @@ in {
+          #        for php. This is currently necessary as the unit runs as root.
+          preStart = ''
+            umask 0077
+            # NOTE: The phpfpm runtime directory is currently preserved
+            #       between restarts.
+            rm -rf /run/phpfpm/nextcloud-credentials/
+            mkdir -p /run/phpfpm/nextcloud-credentials/
+            cp $CREDENTIALS_DIRECTORY/* /run/phpfpm/nextcloud-credentials/
```
Why not symlink them?

networkException replied:
As per my comment above the files are not readable by the nextcloud user.

```
$ pwd
/run/credentials/phpfpm-nextcloud.service
$ ls -lah
total 4.0K
dr-x------  2 root root  60 Dec 22 20:58 .
drwxr-xr-x 11 root root 220 Dec 22 20:58 ..
-r--------  1 root root   7 Dec 22 20:58 dbpass
```

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -85,18 +85,56 @@ let
-  '';
+  # NOTE: The credentials required by all services at runtime, not including things like the
+  #       admin password which is only needed by the setup service.
+  runtimeSystemdCredentials = []
+    ++ (lib.optional (cfg.config.dbpassFile != null) "dbpass:${cfg.config.dbpassFile}")
+    ++ (lib.optional (cfg.config.objectstore.s3.enable) "s3_secret:${cfg.config.objectstore.s3.secretFile}")
+    ++ (lib.optional (cfg.config.objectstore.s3.sseCKeyFile != null) "s3_sse_c_key:${cfg.config.objectstore.s3.sseCKeyFile}");
```
```suggestion
  runtimeSystemdCredentials = lib.optional (cfg.config.dbpassFile != null) "dbpass:${cfg.config.dbpassFile}"
    ++ lib.optional (cfg.config.objectstore.s3.enable) "s3_secret:${cfg.config.objectstore.s3.secretFile}"
    ++ lib.optional (cfg.config.objectstore.s3.sseCKeyFile != null) "s3_sse_c_key:${cfg.config.objectstore.s3.sseCKeyFile}";
```

networkException replied:
I prefer consistent indentation

networkException committed:
nixos/tests/nextcloud: use lib instead of pkgs.lib wherever trivial

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -85,18 +85,56 @@ let
+      #       in the nix_read_secret() php function.
+      #       When there's no CREDENTIALS_DIRECTORY we try to use systemd-run to
+      #       load the credentials just as in a service unit.
+      # NOTE: If there are no credentials that are required at runtime then there's no need
+      #       to load any credentials.
```
IMO we should always execute the script the same way to avoid edge cases and to make testing changes easier.

networkException replied:
If there are no credentials the script would try to use `systemd-run` from within other services, which fails.

It needs some way of knowing if its running in a service or that it shouldn't expect credentials

Ma27 replied:
On a first glance I'd be fine with that, but it would be nice to have a test which runs `occ` from the "outside" to make sure we immediately notice if we're about to break it (that means, a command that fails if secrets are unreadable).

networkException replied:
We have very basic checks for `nextcloud-occ` everywhere: https://github.com/NixOS/nixpkgs/blob/5f81eeb3e39007bcdc3c17488b3225adf2102c84/nixos/tests/nextcloud/default.nix#L70

If you have ideas to expand this let me know

networkException replied:
This is the output of `nextcloud-occ status` with logging enabled for whenever a file path is being read

```diff
       function nix_read_secret($file) {
         $credential_file = getenv("CREDENTIALS_DIRECTORY") . "/" . $file;
+        error_log($credential_file);
         if (!is_readable($credential_file)) {
```

```
$ nextcloud-occ status
/run/credentials/run-u13.service/dbpass
  - installed: true
  - version: 30.0.4.1
  - versionstring: 30.0.4
  - edition: 
  - maintenance: false
  - needsDbUpgrade: false
  - productname: Nextcloud
  - extendedSupport: false
```

I've just pushed additional changes that make the script exit with error code 1, the previously thrown `RuntimeException` would be caught somewhere:

```
subtest: Ensure nextcloud-occ is working
nextcloud: must succeed: nextcloud-occ status
nextcloud # [   23.857755] systemd[1]: Starting /nix/store/b1wvkjx96i3s7wblz38ya0zr8i93zbc5-coreutils-9.5/bin/env NEXTCLOUD_CONFIG_DIR=/var/lib/nextcloud/config /nix/store/c6i258wgrwgg8aladc0v97k39cs2sgp6-php-with-extensions-8.3.15/bin/php occ status...
nextcloud # [   23.897224] systemd[1]: Started /nix/store/b1wvkjx96i3s7wblz38ya0zr8i93zbc5-coreutils-9.5/bin/env NEXTCLOUD_CONFIG_DIR=/var/lib/nextcloud/config /nix/store/c6i258wgrwgg8aladc0v97k39cs2sgp6-php-with-extensions-8.3.15/bin/php occ status.
nextcloud # [   23.984784] nextcloud-update-db-start[1148]: oc_filecache table updated successfully.
nextcloud # [   23.985753] nextcloud-update-db-start[1148]: Adding additional systag_by_objectid index to the oc_systemtag_object_mapping table, this can take some time...
nextcloud # [   24.110756] systemd[1]: run-u4.service: Main process exited, code=exited, status=1/FAILURE
nextcloud # [   24.112105] systemd[1]: run-u4.service: Failed with result 'exit-code'.
nextcloud: output: Cannot read credential 'dbpass' passed by NixOS, $CREDENTIALS_DIRECTORY is not set!

Test "Ensure nextcloud-occ is working" failed with error: "command `nextcloud-occ status` failed (exit code 1)"
```

Ma27 replied:
Looks reasonable, thanks!

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud-notify_push.nix:
```diff
@@ -78,60 +78,73 @@ in
-        RestartSec = "5s";
-        Type = "notify";
+
+      nextcloud-notify_push_setup = {
+        wantedBy = [ "multi-user.target" ];
+        requiredBy = [ "nextcloud-notify_push.service" ];
+        after = [ "nextcloud-notify_push.service" ];
+        serviceConfig = {
+          Type = "oneshot";
+          User = "nextcloud";
+          Group = "nextcloud";
+          ExecStart = "${lib.getExe cfgN.occ} notify_push:setup ${cfg.nextcloudUrl}/push";
+          LoadCredential = config.systemd.services.nextcloud-cron.serviceConfig.LoadCredential;
+        };
       };
```
This will definitely not work. notify_push:setup verifies if the configuration is correct and checks if the domain and trusted proxy settings are correct. It is not similar to nextcloud-setup and must be run after the notify_push service is up. If it fails the service is none functional.

networkException replied:
I'm not too invested in reducing how many credentials the long running process can read. The ordering works just fine like this with `after` and `Type=notify`.

In general having `nextcloud-notify_push_setup.service` compared to `postStart` just changes which service fails in case something goes wrong

Ma27 replied:
> This will definitely not work. notify_push:setup verifies if the configuration is correct and checks if the domain and trusted proxy settings are correct

@SuperSandro2000 I think @networkException is right here and I fail to understand what'll break exactly either.
Please elaborate.

SuperSandro2000 replied:
What do we do if that doesn't work? I can't support systemds complicated credentials options.

Ma27 replied:
@SuperSandro2000 

> Just restarting nextcloud-notify_push_setup won't do anything

`RestartMode=direct` is what you probably need here in combination with `requires`. Then, the dependant units never reach the failed state and get restarted along with the setup unit. That's what I have for essentially all setup units that depend on an external DB server (and thus the first invocation wiill probably fail).

> What do we do if that doesn't work?

If what doesn't work?

> I can't support systemds complicated credentials options.

Your remark is legit, yet solvable.
I fail to see how this is related to the credential feature of systemd.

Ma27 replied:
@networkException I think, the `RestartMode=direct` part is reasonable to do here, isn't it?
Other than that I'd consider the PR good to go (though I'd like to hear back from @SuperSandro2000 first).

Ma27 replied:
ping @SuperSandro2000 what are your thoughts on that suggestion?

SuperSandro2000 replied:
According to https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#RestartMode= `RestartMode=direct` ignores the failure. We don't want to ignore anything. If the occ command fails, things are broken and not working. If that's wrong I don't understand what is going on here.

Ma27 replied:
No it doesn't.

It ignores the failure on a restart, i.e. if a unit gets _restarted_ because of a failure, the failure state is skipped (thus units with a `Require` dependency don't break). However, if the unit fails without getting restarted (e.g. after too many restart), the unit is marked as failed.

To give an example:

```nix
{
  systemd.services = {
    setup = {
      wantedBy = [ "multi-user.target" ];
      script = ''
        exit 1
      '';
      serviceConfig = {
        Restart = "on-failure";
        RestartMode = "direct";
        Type = "oneshot";
      };
    };
    notify = {
      wantedBy = [ "multi-user.target" ];
      requires = [ "setup.service" ];
      after = [ "setup.service" ];
      path = with pkgs; [ coreutils ];
      script = ''
        while true; do
          sleep 1
          echo lalala
        done
      '';
    };
  };
}
```

On activation, the `systemd` log looks like this:

```
Feb 19 18:15:32 snens systemd[1]: Starting setup.service...
Feb 19 18:15:32 snens systemd[1]: setup.service: Main process exited, code=exited, status=1/FAILURE
Feb 19 18:15:32 snens systemd[1]: setup.service: Failed with result 'exit-code'.
Feb 19 18:15:32 snens systemd[1]: setup.service: Scheduled restart job, restart counter is at 1.
Feb 19 18:15:32 snens systemd[1]: Starting setup.service...
Feb 19 18:15:32 snens systemd[1]: setup.service: Main process exited, code=exited, status=1/FAILURE
Feb 19 18:15:32 snens systemd[1]: setup.service: Failed with result 'exit-code'.
Feb 19 18:15:33 snens systemd[1]: setup.service: Scheduled restart job, restart counter is at 2.
Feb 19 18:15:33 snens systemd[1]: Starting setup.service...
Feb 19 18:15:33 snens systemd[1]: setup.service: Main process exited, code=exited, status=1/FAILURE
Feb 19 18:15:33 snens systemd[1]: setup.service: Failed with result 'exit-code'.
Feb 19 18:15:33 snens systemd[1]: setup.service: Scheduled restart job, restart counter is at 3.
Feb 19 18:15:33 snens systemd[1]: Starting setup.service...
Feb 19 18:15:33 snens systemd[1]: setup.service: Main process exited, code=exited, status=1/FAILURE
Feb 19 18:15:33 snens systemd[1]: setup.service: Failed with result 'exit-code'.
Feb 19 18:15:33 snens systemd[1]: setup.service: Scheduled restart job, restart counter is at 4.
Feb 19 18:15:33 snens systemd[1]: Starting setup.service...
Feb 19 18:15:33 snens systemd[1]: setup.service: Main process exited, code=exited, status=1/FAILURE
Feb 19 18:15:33 snens systemd[1]: setup.service: Failed with result 'exit-code'.
Feb 19 18:15:33 snens systemd[1]: setup.service: Scheduled restart job, restart counter is at 5.
Feb 19 18:15:33 snens systemd[1]: setup.service: Start request repeated too quickly.
Feb 19 18:15:33 snens systemd[1]: setup.service: Failed with result 'exit-code'.
Feb 19 18:15:33 snens systemd[1]: Failed to start setup.service.
Feb 19 18:15:33 snens systemd[1]: Dependency failed for notify.service.
Feb 19 18:15:33 snens systemd[1]: notify.service: Job notify.service/start failed with result 'dependency'.
```

As you can see, `notify.service` will fail _after_ `setup.service` failed to start 5 times (due to too many restarts).
And yes, `setup.service` is marked as failed.

Now, with the `RestartMode=direct` commented out, `notify.service` will fail (with result 'dependency') after the first failure of `setup.service`. So yes, `RestartMode=direct` is what we want.

networkException replied:
Thank you

Ma27 commented:
Thanks!
Will try to review in the upcoming days.

networkException committed:
nixos/tests/nextcloud: fix redis cache non empty tests

This patch changes the implementation of the subtests to
check for redis' cache being non empty to only run redis-cli
and jq in a shell and assert the returned length in python.

This fixes jq "len" simply not compiling and makes sure
regressions get noticed.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -1064,6 +1111,21 @@ in {
+            umask 0077
+            # NOTE: The phpfpm runtime directory is currently preserved
+            #       between restarts.
+            rm -rf /run/phpfpm/nextcloud-credentials/
```
Also... doesn't this mean that other phpfpm instances can see these values (even though only readable as root, so not from PHP code)?

networkException replied:
> I think it would be nicer to use `$RUNTIME_DIRECTORY/nextcloud-credentials/` here.

I don't think this currently applies here but `$RUNTIME_DIRECTORY` has the pesky detail that you can technically specify multiple directories which will end up as `/run/a:/run/b` in the variable

networkException replied:
> Also... doesn't this mean that other phpfpm instances can see these values (even though only readable as root, so not from PHP code)?

Yes sadly thats the case, briefly as `root` and then as `nextcloud`. I don't know phpfpm good enough to know what would happen if the service were rootless, but that would be much cleaner here

Ma27 replied:
> I don't think this currently applies here but $RUNTIME_DIRECTORY has the pesky detail that you can technically specify multiple directories which will end up as /run/a:/run/b in the variable

Meh, makes perfect sense for a variable name in signular :/ 
but good point, I'll retract that.

> Yes sadly thats the case, briefly as root and then as nextcloud. I don't know phpfpm good enough to know what would happen if the service were rootless, but that would be much cleaner here

My main problem is that the runtime directory is used for each phpfpm service, i.e. each PHP application on the same host will have this in its runtime directory (sure, different user then, but this smells like a disaster waiting to happen).

But given that we can specify multiple runtime directories, wouldn't it be an option to add another directory to the phpfpm-nextcloud service and place the secrets into that dir?

networkException replied:
> But given that we can specify multiple runtime directories, wouldn't it be an option to add another directory to the phpfpm-nextcloud service and place the secrets into that dir?

Sure, good point. I'll work on this in a bit

networkException replied:
I've added a separate runtime directory to the service (https://github.com/NixOS/nixpkgs/compare/dedb0ac10499c7406ce55399db5ac7fdf3568907..4d4d3672a52c630e4a3f7970b5c3aa1bb2beda00). Note that sadly `RuntimeDirectoryPreserve=true` as well as the mode also applies. I also don't know a way to avoid the `lib.mkForce` here.

Ma27 replied:
> I also don't know a way to avoid the lib.mkForce here.

I think you'd have to turn `RuntimeDirectory` into a list in the phpfpm module. But I'm also fine with the `mkForce` here.

> Note that sadly RuntimeDirectoryPreserve=true as well as the mode also applies

`/run` is a tmpfs anyways, so it's not persisted across reboots. This still sucks, but at least we have a dedicated runtime directory for this, so this doesn't get leaked to other services, so I'm fine with this solution.

Resolving.

networkException replied:
> I think you'd have to turn `RuntimeDirectory` into a list in the phpfpm module. But I'm also fine with the `mkForce` here.

I think I'd still need to force override the directive because appending to the list isn't really possible here. I could add a module option to phpfpm maybe?

networkException commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -1020,19 +1052,29 @@ in {
         nextcloud-cron = {
           after = [ "nextcloud-setup.service" ];
+          # FIXME: In contrast to the occ wrapper script running phpCli directly will not
+          #        set NEXTCLOUD_CONFIG_DIR correctly currently.
```
Maybe a note is better, I added this after removing the explicit (somewhat duplicate) variable from the `nextcloud-update-db` service

Ma27 replied:
I think the comment is good for clarification, but I don't really see this as something super-actionable, so I think I'd prefer `NOTE:` here.

networkException replied:
Changed in https://github.com/NixOS/nixpkgs/compare/db42190f5e64dd4c8ce24b138d1195b3e307b1e3..e08ac4cf2a450205b8c7484139cb2d6989d16936

networkException commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -143,16 +185,26 @@ let
-            $file
+      function nix_read_secret($credential_name) {
+        $credentials_directory = getenv("CREDENTIALS_DIRECTORY");
+        if (strlen($credentials_directory) === 0) {
```
Sure, I'm not really familiar with the language and the internet was a bit weary about the bool cast matching for more than just empty string.

networkException replied:
Changed that

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -1009,57 +1039,94 @@ in {
+            # NOTE: This early returns the script when nextcloud is in maintenance mode
+            #       or needs `occ upgrade`. Using ExecCondition= is not possible here
+            #       because it doesn't work with systemd credentials.
+            ${lib.getExe occ} status -e; [ $? -ne 0 ] && exit 0
```
Hmm, doesn't this fail the service since `script` adds a `set -e`?

networkException replied:
I've changed the implementation to use `status --output=json`, feels nicer compared to `-e` anyways: https://github.com/NixOS/nixpkgs/compare/4d4d3672a52c630e4a3f7970b5c3aa1bb2beda00..3d64615e8a60490360283fad2ed7b8bca6f29372

Ma27 replied:
Just verified by hand that this behaves as expected.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -1009,57 +1039,94 @@ in {
+          script = ''
+            # NOTE: This early returns the script when nextcloud is in maintenance mode
+            #       or needs `occ upgrade`. Using ExecCondition= is not possible here
+            #       because it doesn't work with systemd credentials.
```
fwiw opened https://github.com/systemd/systemd/issues/35788 to discuss whether we could change this upstream.

Ma27 replied:
Resolving. You successfully sniped me into trying this, I just need to find a bit of time and regardless of that, this shouldn't block this change.

networkException commented:
Rebased onto `master` and checked that the test still pass

Ma27 commented:
So the code mostly looks reasonable to me and the tests are passing.
I guess I'll pick it to my personal nixpkgs branch, deploy my private instance and see how it turns out.

Ma27 reviewed (commented):
Works as expected, thanks :+1: 

No approval yet though because

* I think a brief entry in the release notes makes sense (I'm probably not the only one with a heavily configured Nextcloud and I had to adjust some parts of my config. Doesn't need much detail in this case, but I'd like to see it mentioned nonetheless)
* The notify_push thing I don't know much abotu is still unresolved (https://github.com/NixOS/nixpkgs/pull/367433/files#r1895052681): cc @SuperSandro2000 @dasj @helsinki-Jo 

Also cc @britter @dotlambda you may also be interested in taking a look.

networkException commented:
Rebased onto `master` again, test still pass

networkException commented:
> I had to adjust some parts of my config

What did you need to change?

Ma27 commented:
> What did you need to change?

`adminpassFile` was set to a `/run/credentials/...` path already and this broke `nextcloud-setup` in the credentials setup phase.

Also, I had another service for generating previews for new images as a systemd timer and inherited `ExecCOndition` from nextcloud-setup which broke evaluation.

networkException committed:
nixos/nextcloud: document systemd credentials as a backwards incompatible change in the 25.05 release notes

This patch adds a release note entry to the 25.05 release
about the use of systemd credentials to read in secrets.

It's part of the backward incompatibilities section as
changes to the behavior of `nextcloud-occ` might break
existing scripts.

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud-notify_push.nix:
```diff
@@ -78,60 +78,73 @@ in
-        RestartSec = "5s";
-        Type = "notify";
+
+      nextcloud-notify_push_setup = {
```
```suggestion
      nextcloud-notify_push-verify= {
```
We are not really setting up anything, the command is just named misleading.

Ma27 replied:
Uhm... I think it's reasonable to call the service `setup` if the _setup_ command of notify_push is invoked ;)

SuperSandro2000 replied:
Well, yeah, but we don't do what the setup command usually does. Normally it guides you through configuring and installing notify (the nixos module did all of that already) and as a last step it verifies if things are working. Since we did all the setup steps already, we skip all of that and only use the last part. 😅 So we are kinda abusing the command.

Ma27 replied:
It seems as if that also modifies the app's configuration: https://github.com/nextcloud/notify_push/blob/main/lib/Command/Setup.php#L51-L52

The NEXTCLOUD_URL seems to eb read by the Rust part only.
In fact, when replacing this with `${coreutils}/bin/true`, I get

```
client # Error:
client #   × notify_push app doesn't seem to be enabled or setup
```

whilst running the test.

Ma27 committed:
nixos/nextcloud-notify_push: use RestartMode=direct

`nextcloud-notify_push.service` requires
`nextcloud-notify_push-setup.service`. If the latter fails (e.g. because
of Nextcloud not being there yet), the push service would also fail with
result 'dependency'.

RestartMode=direct doesn't put a unit into failed state IF it's about to
be restarted again. That way, `nextcloud-notify_push` will await several
restart attempts. Only if the unit fails due to a rate-limit (i.e. too
many restarts), the push service will also fail.

If the startup is still too slow, it may make sense for administrators to
configure higher intervals between the start attempts with RestartSec.

Ma27 reviewed (approved):
I deploy my Nextcloud with these patches on top for ~2 months now. I'm getting 'eepy now, so I'll merge tomorrow afternoon.

If people still have reservations about this, please speak up.

SuperSandro2000 commented:
I just cherry-picked that into my local nixos-unstable fork and shellcheck fails. I will do a follow up PR that fixes that.

SuperSandro2000 commented:
#387913

Ma27 commented:
> I just cherry-picked that into my local nixos-unstable fork and shellcheck fails. I will do a follow up PR that fixes that.

Where is the shellcheck even running? Tests were all building fwiw.

SuperSandro2000 commented:
I have systemd.enableStrictShellChecks
 https://search.nixos.org/options?channel=unstable&show=systemd.enableStrictShellChecks&from=0&size=50&sort=relevance&type=packages&query=shellcheck enabled globally, so it didn't trigger when you build it.

Ma27 commented:
Oof... that sounds like an invitation for random build failures whenever one updates their systems.
To be clear: I've seen the missing quotes, but I decided to not start another feedback cycle for that given that (1) it is __guaranteed__ that CREDENTIALS_DIRECTORY exists and (2) the quotes do not improve the error reporting if it it doesn't.

I mean, I'll merge since it doesn't hurt and will fix people's setup with this option, but... I'm not sure if this option is a good idea in the first place.

SuperSandro2000 commented:
I have it on since a while and after the initial fixes I barely get any regressions. Maybe every 2 to 4 weeks. It is surprising. 😂
